src/test_fixture/ipa.rs

```diff
@@ -72,29 +72,47 @@ pub fn generate_random_user_records_in_reverse_chronological_order(
         });
     }
 
-    // sort in reverse time order
-    records_for_user.sort_unstable_by(|a, b| b.timestamp.cmp(&a.timestamp));
+    // sort in time order
+    records_for_user.sort_unstable_by(|a, b| a.timestamp.cmp(&b.timestamp));
 
     records_for_user
 }
 
-/// Assumes records all belong to the same user, and are in reverse chronological order
+/// Assumes records all belong to the same user, and are in chronological order
 /// Will give incorrect results if this is not true
 #[allow(clippy::missing_panics_doc)]
 pub fn update_expected_output_for_user(
     records_for_user: &[TestRawDataRecord],
     expected_results: &mut [u32],
     per_user_cap: u32,
-    _attribution_window_seconds: u32, // TODO(taikiy): compute the output with the attribution window
+    attribution_window_seconds: u32,
 ) {
+    // Calculate time deltas from the nearest source report to the current trigger report.
+    let time_delta = records_for_user
+        .iter()
+        .scan(0, |source_ts, x| {
+            if x.is_trigger_report {
+                Some(x.timestamp - *source_ts)
+            } else {
+                *source_ts = x.timestamp;
+                Some(0)
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // reverse the records so we can iterate from the most recent to the oldest
+    let mut reversed_records = records_for_user.to_vec();
+    reversed_records.reverse();
+
     let mut pending_trigger_value = 0;
```

Comment:
I don't want to do it this way. I do not like the way that this approach duplicates the logic of searching backwards to find the last-touch. I forsee difficulties when we attempt to extend this to multi-touch attribution, because it will force us to make changes in two places.

I would rather keep it as just a single pass over the data (not two), with no need to reverse the list.

I recommend the following:

1. Replace `pending_trigger_value` with a Vec of `pending_trigger_reports`, and just push trigger reports into it.
2. When encountering a source event, instead of checking if `pending_trigger_value > 0` just check if `pending_trigger_reports.len() > 0`
3. Then, iterate through each of the `pending_trigger_reports`, (for which you'll now have individual timestamps) and compute the delta to this particular source event. If it's longer than the window, just continue. Otherwise, do the current logic.

Reply:
Updated!